### ISSUE TITLE
XDMFFile does not close correctly at end of python session

### ISSUE THREAD

**Issue by nate-sime** [FEniCS/dolfinx#75]
Consider the following
```
from dolfin import *

mesh = UnitSquareMesh(MPI.comm_world, 1, 1)
mesh.geometry.coord_mapping = fem.create_coordinate_map(mesh.ufl_domain())

V = FunctionSpace(mesh, "CG", 1)

uu = Function(V)
XDMFFile("out1.xdmf").write(uu, float(100.0), XDMFFile.Encoding.HDF5)

xdmf2 = XDMFFile("out2.xdmf")
xdmf2.write(uu, float(100.0), XDMFFile.Encoding.HDF5)
```
Hopefully the error is reproducible. I see that out1.xdmf will be read fine in paraview, whilst out2.xdmf has a corrupt HDF5 file.

I can fix the problem by adding XDMF::close to the python layer and explicitly calling `xdmf2.close()` at the end of the code.

**Comment by garth-wells** [comment]
```
with XDMFFile("out2.xdmf") as f:
   f.write(uu, float(100.0), XDMFFile.Encoding.HDF5)
```
is the recommended approach. Garbage is collection is complicated, especially when calling 3rd party libs, so we should change the i/o syntax in demos.

**Comment by michalhabera** [comment]
As @garth-wells says, context manager and python's GC is recommended approach. However is not possible for the old `XDMFFile::write` unless you want to have your time loop wrapped and tabbed in context manager. That being said, once you destroy XDMFFile object the old `write()` method won't append it. On the other hand new `write_checkpoint` appends it and "local" context managers are possible for time loops.

There is also a parameter `flush_output` (https://github.com/FEniCS/dolfinx/blob/master/cpp/dolfin/io/XDMFFile.cpp#L58) which forces underlying HDF file to be closed after every `XDMFFile::write` method call. That should solve the issue you are reporting. 
But I do not like these parameters, especially when they are set to an object as whole (e.g. XDMFFile) but they apply only to part of that object (like XDMFFile::write on functions). Very confusing.

If you'd like to do very quick fix of this I'd recommend putting flush_output = True by default.

**Comment by garth-wells** [comment]
@michalhabera @nate-sime I would suggest using `with` where possible, and in other cases use `close()`. We should update demos to follow this pattern.

**Comment by nate-sime** [comment]
Are there any objections to having `XDMF::close` back in the python interface? I agree with using `with` wherever possible, but the nested loops for complicated FE problems could quickly get out of hand...

**Comment by chrisrichardson** [comment]
You can always use `del xdmf2`. But probably `close` is neater

**Comment by garth-wells** [comment]
@nate-sime is this now resolved?

**Comment by nate-sime** [comment]
Issues addressed and resolved in PR #77
